envoy_qatzip/qatzip_filter.cc

```diff
@@ -54,10 +65,32 @@ unsigned int QatzipFilterConfig::streamBufferSizeUint(Protobuf::uint32 stream_bu
 }
 
 std::unique_ptr<Compressor::Compressor> QatzipFilterConfig::makeCompressor() {
-  auto compressor = std::make_unique<Compressor::QatzipCompressorImpl>();
-  compressor->init(compressionLevel(), hardwareBufferSize(), streamBufferSize(),
-                   inputSizeThreshold());
-  return compressor;
+  return std::make_unique<Compressor::QatzipCompressorImpl>(
+      tls_slot_->getTyped<QatzipThreadLocal>().getSession());
+}
+
+QatzipFilterConfig::QatzipThreadLocal::QatzipThreadLocal(QzSessionParams_T params)
+    : params_(params), session_{}, initialized_(false) {}
+
+QatzipFilterConfig::QatzipThreadLocal::~QatzipThreadLocal() {
+  if (initialized_) {
+    qzTeardownSession(&session_);
+    qzClose(&session_);
+  }
+}
+
+QzSession_T* QatzipFilterConfig::QatzipThreadLocal::getSession() {
+  // The session must be initialized only once in every worker thread.
+  if (!initialized_) {
+
+    int status = qzInit(&session_, params_.sw_backup);
+    RELEASE_ASSERT(status == QZ_OK || status == QZ_DUPLICATE, "failed to initialize hardware");
```

Comment:
If the session must be initialized only once, isn't `QZ_DUPLICATE` an error case?

Reply:
The API doc says it's not an error but only a sign that the function was not processed